man/systemd-timesyncd.service.xml

```diff
@@ -78,7 +78,8 @@
           <para>The modification time ("mtime") of this file is updated on each successful NTP synchronization
           or after each <varname>SaveIntervalSec=</varname> time interval, as specified in
           <citerefentry><refentrytitle>timesyncd.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
-          At the minimum, it will be set to the systemd build date. It is used to ensure that the system clock
+          At the minimum, it will be set to the systemd build date (specified as modification time ("mtime") of
+          <filename>/usr/lib/clock-epoch</filename>). It is used to ensure that the system clock
```

Comment:
@darkdragon-001 ?

Reply:
I think I still don't understand how it works...

@poettering So the `mtime` of `/var/lib/systemd/timesync/clock` is used if it exists, otherwise the `mtime` of `/usr/lib/clock-epoch` is used if it exists and if both don't then the build time is used? If so where is the build time stored?

Reply:
pid1 early on uses /usr/lib/clock-epoch's mtime if the file exists. if it does not, it uses the built-in time epoch (which is typically derived from the NEWS file timestamp ad build-time, but there are other options)

systemd-timesyncd then much later on might bump again via /var/lib/systemd/timesync/clock's mtime.

the stuff pid 1 does is a complete baseline concept, it is based on static timestamps that are never updated during runtime.

the stuff timesyncd is much more useful, since it ensures rough monotonicity of the clock. However it can do this only much later during boot, at the point where /var/ becomes accessible. Since this logic actually maintains a dynamic timestamp it's a much more powerful concept.

So, pid 1 clock bump is early and static, and timesyncd clock bump is late and dynamic, if you so will.

Reply:
Thanks for the explanation! I gave it another try. Can you have a look if it is not accurate enough?
If not, could you please provide some draft?